### PR TITLE
Tutorial smart contract guest book

### PR DESCRIPTION
**Pull request by bucanero**
This PR is based on (#675)

### CODE REVIEW

File: docs/tutorials/contracts/guest-book.md [hunk 1]
```diff
@@ -0,0 +1,993 @@
...
+
+You need `near-cli` installed globally. Here's how:
+
+    npm install --global near-cli
+
+This will give you the `near` [CLI] tool. Ensure that it's installed with:
+
+    near --version
+
+Next, clone the repo [here]. We will start with the `boilerplate` branch, and slowly progress towards the master branch with the completed code.
```
Comment (thisisjoshford) [inline docs/tutorials/contracts/guest-book.md:24]: need a link to the repo

File: docs/tutorials/contracts/guest-book.md [hunk 2]
```diff
@@ -0,0 +1,993 @@
...
+
+This will give you the `near` [CLI] tool. Ensure that it's installed with:
+
+    near --version
+
+Next, clone the repo [here]. We will start with the `boilerplate` branch, and slowly progress towards the master branch with the completed code. 
+
+Alternatively, create a basic NEAR React project with package [create-near-app] using the command:
+
+    yarn create near-app path/to/my-near-react-app
```
Comment (thisisjoshford) [inline docs/tutorials/contracts/guest-book.md:28]:
```suggestion
    npx create-near-app YOUR_AWESOME_APP --frontend=react
```

File: docs/tutorials/contracts/guest-book.md [hunk 3]
```diff
@@ -0,0 +1,993 @@
...
+
+Just adjust the folder structure to match the branch we will be referencing.
```
Comment (thisisjoshford) [inline docs/tutorials/contracts/guest-book.md:30]:
```suggestion
```

File: docs/tutorials/contracts/guest-book.md [hunk 4]
```diff
@@ -0,0 +1,993 @@
...
+### Local Setup
+
+To run this project locally:
+
+1. Prerequisites: Make sure you have [Node.js](https://nodejs.org) ≥ 12 installed, then use it to install [yarn]: `npm install --global yarn` (or just `npm i -g yarn`)
+2. Install dependencies: `yarn install` (or just `yarn`)
+3. Run the local development server: `yarn dev` (see `package.json` for a
+   full list of `scripts` you can run with `yarn`)
+
+Now you'll have a local development environment backed by the NEAR _TestNet_. Running `yarn dev` will tell you the URL you can visit in your browser to see the app.
```
Comment (thisisjoshford) [inline docs/tutorials/contracts/guest-book.md:42]:
```suggestion
Now you'll have a local development environment connected to NEAR's _testnet_ network. Running `yarn dev` will tell you the URL you can visit in your browser to see the app.
```

File: docs/tutorials/contracts/guest-book.md [hunk 5]
```diff
@@ -0,0 +1,993 @@
...
+
+
+### Exploring The Code
+
+1. The backend code (smart contract) lives in the `/assembly` folder. This code deploys to
+   the NEAR blockchain when you run `yarn deploy:contract`  – [learn more
+   about NEAR smart contracts][smart contract docs].
```
Comment (thisisjoshford) [inline docs/tutorials/contracts/guest-book.md:49]: need url

File: docs/tutorials/contracts/guest-book.md [hunk 6]
```diff
@@ -0,0 +1,993 @@
...
+
+Both contract and client-side code will auto-reload as you update source files.
+
+
+
+## Contract
+
+If we imagine the blockchain to be a sort of database, then smart contracts function like an ORM or API with read/write access to the blockchain. 
+
+The methods that do _NOT_ update the state of your application are called _view_ methods. Those that do are called _call_ or _change_ methods.
```
Comment (thisisjoshford) [inline docs/tutorials/contracts/guest-book.md:65]:
```suggestion
The methods that do _NOT_ update the state of your application are called _view_ methods. Those that do modify state are called _change_ methods.
```
You can actually just view state with a contract call. Why someone would actually do that is beyond me lol

File: docs/tutorials/contracts/guest-book.md [hunk 7]
```diff
@@ -0,0 +1,993 @@
...
+
+
+## Contract: Develop 
+
+For this section we will be working in the `assembly/` folder of your project, and all code is written in TypeScript.
```
Comment (thisisjoshford) [inline docs/tutorials/contracts/guest-book.md:70]:
```suggestion
For this section we will be working in the `assembly/` folder of your project, and all code is written in AssemblyScript.
```

Its important that we make sure we distinguish TS from AS as they may seem similar they are very different.

File: docs/tutorials/contracts/guest-book.md [hunk 8]
```diff
@@ -0,0 +1,993 @@
...
+Well done! We created a smart contract and tested it. 
+
+Write a few more simple tests of your own, and see if we missed any edge cases.
+
+Otherwise, let's deploy our smart contract and interact with it from the terminal.
+
+
+## Contract: Deploy
+
+We deploy our smart contract to NEAR _TestNet_, with either an auto-generated account id (dev deploy) or a permanent one we've already set up.
```
Comment (thisisjoshford) [inline docs/tutorials/contracts/guest-book.md:377]:
```suggestion
We deploy our smart contract to NEAR's `testnet` network, with either an auto-generated account id ([dev-deploy](https://docs.near.org/docs/tools/near-cli#near-dev-deploy)) or an existing account id you already created.
```

File: docs/tutorials/contracts/guest-book.md [hunk 9]
```diff
@@ -0,0 +1,993 @@
...
+
+
+## Contract: Deploy
+
+We deploy our smart contract to NEAR _TestNet_, with either an auto-generated account id (dev deploy) or a permanent one we've already set up. 
+
+
+### Dev Deploy
+
+Every smart contract in NEAR has its [own associated account][NEAR accounts].
```
Comment (thisisjoshford) [inline docs/tutorials/contracts/guest-book.md:382]: add url :)

File: docs/tutorials/contracts/guest-book.md [hunk 10]
```diff
@@ -0,0 +1,993 @@
...
+
+```
+  yarn dev
+```
+
+The above deploys your smart contract to NEAR _TestNet_ with a throwaway account. It also starts a dev server, which we will need when we work on our front end.
```
Comment (thisisjoshford) [inline docs/tutorials/contracts/guest-book.md:388]:
```suggestion
The above deploys your smart contract to `testnet` with a [dev-account](https://docs.near.org/docs/concepts/account#dev-accounts). It also starts a local server, which we will need when we work on our front end. 
```

File: docs/tutorials/contracts/guest-book.md [hunk 11]
```diff
@@ -0,0 +1,993 @@
...
+
+If you want to start interacting with your contracts without starting a local server, you can simply run:
+
+```
+  near dev-deploy
+```
+
+It does pretty much the same thing as `yarn dev`, but without the dev server.  So it will generate an account for your contract, `dev-some-autogenerated-account`.
+
+If you have already configured permanent account for your contract, you can deploy it with the following command:
```
Comment (thisisjoshford) [inline docs/tutorials/contracts/guest-book.md:398]:
```suggestion
If you have already created an account for your contract, you can deploy it with the following command: 
```

File: docs/tutorials/contracts/guest-book.md [hunk 12]
```diff
@@ -0,0 +1,993 @@
...
+It does pretty much the same thing as `yarn dev`, but without the dev server.  So it will generate an account for your contract, `dev-some-autogenerated-account`.
+
+If you have already configured permanent account for your contract, you can deploy it with the following command: 
+
+
+### Prod Deploy 
+
+``` 
+  near deploy \
+  --contractName=permanent-contract-accountid-you-set-up.testnet \
```
Comment (thisisjoshford) [inline docs/tutorials/contracts/guest-book.md:405]:
```suggestion
  --contractName=YOUR_ACCOUNT_ID.testnet \
```

File: docs/tutorials/contracts/guest-book.md [hunk 13]
```diff
@@ -0,0 +1,993 @@
...
+  --keyPath=./path-to/testnet/permanent-contract-accountid-you-set-up.testnet.json \
```
Comment (thisisjoshford) [inline docs/tutorials/contracts/guest-book.md:406]:
```suggestion
  --keyPath=./path-to/testnet/YOUR_ACCOUNT_ID.testnet.json \
```

File: docs/tutorials/contracts/guest-book.md [hunk 14]
```diff
@@ -0,0 +1,993 @@
...
+```
+
+
+### Invoking Contract Methods
+
+
+The following command calls our main change method, `addMessage`.
+
+```
+ near call dev-some-autogenerated-account addMessage '{"text": "hello guestbook"}' --accountId someValidNEARAccount.testnet
```
Comment (thisisjoshford) [inline docs/tutorials/contracts/guest-book.md:417]:
```suggestion
 near call YOUR_ACCOUNT_ID.testnet addMessage '{"text": "hello guestbook"}' --accountId AN_ACCOUNT_YOU_OWN.testnet
```

let's also add a note here about how for the `--accountId` flag the `AN_ACCOUNT_YOU_OWN.testnet` needs to have credentials stored locally using `near login`

File: docs/tutorials/contracts/guest-book.md [hunk 15]
```diff
@@ -0,0 +1,993 @@
...
+```
+
+then run:
+
+```
+ near call dev-some-autogenerated-account getMessages --accountId someValidNEARAccount.testnet
```
Comment (thisisjoshford) [inline docs/tutorials/contracts/guest-book.md:423]:
```suggestion
 near call YOUR_ACCOUNT_ID.testnet getMessages --accountId AN_ACCOUNT_YOU_OWN.testnet
```

File: docs/tutorials/contracts/guest-book.md [hunk 16]
```diff
@@ -0,0 +1,993 @@
...
+```
+
+Your terminal output should resemble this:
+
+```
+[
+  {
+    premium: false,
+    sender: 'someValidNEARAccount.testnet',
```
Comment (thisisjoshford) [inline docs/tutorials/contracts/guest-book.md:432]:
```suggestion
    sender: 'AN_ACCOUNT_YOU_OWN.testnet',
```

File: docs/tutorials/contracts/guest-book.md [hunk 17]
```diff
@@ -0,0 +1,993 @@
...
+    text: 'hello guestbook'
+  }
+]
+```
+
+Fantastic! Our backend is ready to send and receive. Let's build our UI!
+
+### Switching to a permanent account
+
+When you're ready to switch over from a dev account to a permanent one, here's how:
```
Comment (thisisjoshford) [inline docs/tutorials/contracts/guest-book.md:442]:
```suggestion
When you're ready to switch over from a dev account to a custom one, here's how:
```

File: docs/tutorials/contracts/guest-book.md [hunk 18]
```diff
@@ -0,0 +1,993 @@
...
+
+##### Step 1: Create an account for the contract
+
+Visit [NEAR Wallet] and make a new account. You'll be deploying these smart contracts to this new account.
```
Comment (thisisjoshford) [inline docs/tutorials/contracts/guest-book.md:446]:
```suggestion
Visit [NEAR Wallet](https://wallet.testnet.near.org/) and make a new account. You'll be deploying these smart contracts to this new account.
```

File: docs/tutorials/contracts/guest-book.md [hunk 19]
```diff
@@ -0,0 +1,993 @@
...
+Now authorize NEAR CLI for this new account, and follow the instructions it gives you:
```
Comment (thisisjoshford) [inline docs/tutorials/contracts/guest-book.md:447]: Explain what `near login` does

File: docs/tutorials/contracts/guest-book.md [hunk 20]
```diff
@@ -0,0 +1,993 @@
...
+
+
+## Front-End
+
+In this section we will not only build (and test) a way for the user to interact with our contract in the browser, we will also allow them to go premium using their NEAR wallet. 
+
+
+## Front-End: Develop
+
+Before we begin, if you are new to [React], you may want to spend some time reviewing the docs before proceeding. This tutorial focuses on ramping up your smart contract skills, so we won't be spending too much time on the intricacies of MVC, etc. Also, if you are not familiar with _functional (stateless) components_ and _hooks_, definitely, take some time to get acquainted with them.
```
Comment (thisisjoshford) [inline docs/tutorials/contracts/guest-book.md:483]: add react url

File: docs/tutorials/contracts/guest-book.md [hunk 21]
```diff
@@ -0,0 +1,993 @@
...
+* User can log in/out.
+* User can submit a message.
+* Messages are retrieved once a user has logged in.
+* User can add a donation with their message.
+
+That's it! Go ahead and run `yarn dev` in the terminal. You should see something like this:
+
+![Guest Book Logged In](/docs/assets/guest-book_fe_logged-in.jpg)
+
+Try logging out then back in again. NEAR explorer will check for your wallet credentials stored locally in the browser, and will ask you to authorize this app. You can try some of the other features our app offers, and it will all happen in our dev environment.
```
Comment (thisisjoshford) [inline docs/tutorials/contracts/guest-book.md:713]:
```suggestion
Try logging out then back in again. NEAR Wallet will check for your wallet credentials stored locally in the browser, and will ask you to authorize this app.
```

**Review by thisisjoshford — CHANGES_REQUESTED** [review]
Left some comments

File: docs/tutorials/contracts/guest-book.md (hunk 1 above)
Comment (bucanero) [reply docs/tutorials/contracts/guest-book.md:24]: Fixed (links at the bottom)

File: docs/tutorials/contracts/guest-book.md (hunk 5 above)
Comment (bucanero) [reply docs/tutorials/contracts/guest-book.md:48]: Fixed

File: docs/tutorials/contracts/guest-book.md (hunk 20 above)
Comment (bucanero) [reply docs/tutorials/contracts/guest-book.md:482]: Fixed

File: docs/tutorials/contracts/guest-book.md (hunk 9 above)
Comment (bucanero) [reply docs/tutorials/contracts/guest-book.md:381]: Fixed

File: docs/tutorials/contracts/guest-book.md (hunk 14 above)
Comment (bucanero) [reply docs/tutorials/contracts/guest-book.md]: Fixed in a030ab61703335d8aa795892ca6dd80473a0a40b

File: docs/tutorials/contracts/guest-book.md (hunk 19 above)
Comment (bucanero) [reply docs/tutorials/contracts/guest-book.md:448]: Fixed in a030ab61703335d8aa795892ca6dd80473a0a40b

File: docs/tutorials/contracts/guest-book.md [hunk 22]
```diff
@@ -0,0 +1,994 @@
...
+## Contract
+
+If we imagine the blockchain to be a sort of database, then smart contracts function like an ORM or API with read/write access to the blockchain. 
+
+The methods that do _NOT_ update the state of your application are called _view_ methods. Those that do modify state are called _change_ methods.
+
+
+## Contract: Develop 
+
+We can start by checking out the `boilerplate` branch and navigating to the `assembly/` folder of the project where all our AssemblyScript code is written. If you want the completed code for this section only, see the `contract-test` branch. If you want the fully completed code, checkout the `main` branch.
```
Comment (bucanero) [inline docs/tutorials/contracts/guest-book.md:63]:
```suggestion
We can start by checking out the `boilerplate` branch and navigating to the project's `assembly/` folder where all our AssemblyScript code is written. If you want the completed code for this section only, see the `contract-test` branch. If you want the fully completed code, checkout the `main` branch.   
```

**Review by BenKurrek — APPROVED** [review]

File: docs/tutorials/contracts/guest-book.md [hunk 23]
```diff
@@ -0,0 +1,996 @@
...
+
+```
+  yarn test
+```
+
+Once finished, you should see passing tests that look like the following:
+
+![Guest Book Contract Unit Test](/docs/assets/guest-book_as_unit-test.jpg)
+
+> If you get an error, make sure you have node version 14.16 installed.
```
Comment (bucanero) [inline docs/tutorials/contracts/guest-book.md:367]:
```suggestion
> **Note:** if you get an error, make sure you have Node.js version 14.16 installed.
```
